### PR TITLE
[Core] Fix recursive cancelation crashes the worker when actor task is a child. 

### LINKED ISSUE #31398: [Core] Consistently reproducible WorkerCrashedError: task_spec.cc:200:  Check failed: sched_cls_id_ > 0
### What happened + What you expected to happen

The reproduction script below reliably trigger WorkerCrashedError on my environment. The error was:

```
$ python ray_crash_bug.py 
2023-01-03 10:33:58,086 INFO worker.py:1538 -- Started a local Ray instance.
2023-01-03 10:33:59,262 WARNING worker.py:1851 -- A worker died or was killed while executing a task by an unexpected system error. To troubleshoot the problem, check the logs for the dead worker. RayTask ID: be0b7347b0de0ee661f8264c798330a1fb9495ac01000000 Worker ID: 3b61456135c8a2927850fed0656f24db22c3f18b7709071caf6110cd Node ID: bb7ad71d7c2e3dc14ef6d5f1d815e38190100fea0a70c30cac2c1f75 Worker IP address: x.x.x.x Worker port: 40371 Worker PID: 279406 Worker exit type: SYSTEM_ERROR Worker exit detail: Worker unexpectedly exits with a connection error code 2. End of file. There are some potential root causes. (1) The process is killed by SIGKILL by OOM killer due to high memory usage. (2) ray stop --force is called. (3) The worker is crashed unexpectedly due to SIGSEGV or other unexpected errors.
Traceback (most recent call last):
  File "/path/to/my/project/ray_crash_bug.py", line 36, in <module>
    ray.get(t)
  File "/path/to/my/project/.venv/lib/python3.10/site-packages/ray/_private/client_mode_hook.py", line 105, in wrapper
    return func(*args, **kwargs)
  File "/path/to/my/project/.venv/lib/python3.10/site-packages/ray/_private/worker.py", line 2311, in get
    raise value
ray.exceptions.WorkerCrashedError: The worker died unexpectedly while executing this task. Check python-core-worker-*.log files for more information.
(task pid=279406) [2023-01-03 10:33:59,257 C 279406 279759] task_spec.cc:200:  Check failed: sched_cls_id_ > 0 
(task pid=279406) *** StackTrace Information ***
(task pid=279406) /path/to/my/project/.venv/lib/python3.10/site-packages/ray/_raylet.so(+0xce4b2a) [0x7fd24d343b2a] ray::operator<<()
(task pid=279406) /path/to/my/project/.venv/lib/python3.10/site-packages/ray/_raylet.so(+0xce6612) [0x7fd24d345612] ray::SpdLogMessage::Flush()
(task pid=279406) /path/to/my/project/.venv/lib/python3.10/site-packages/ray/_raylet.so(_ZN3ray6RayLogD1Ev+0x37) [0x7fd24d345927] ray::RayLog::~RayLog()
(task pid=279406) /path/to/my/project/.venv/lib/python3.10/site-packages/ray/_raylet.so(+0x86ca25) [0x7fd24cecba25] ray::TaskSpecification::GetSchedulingClass()
(task pid=279406) /path/to/my/project/.venv/lib/python3.10/site-packages/ray/_raylet.so(_ZN3ray4core29CoreWorkerDirectTaskSubmitter10CancelTaskENS_17TaskSpecificationEbb+0x1dd) [0x7fd24ccab11d] ray::core::CoreWorkerDirectTaskSubmitter::CancelTask()
(task pid=279406) /path/to/my/project/.venv/lib/python3.10/site-packages/ray/_raylet.so(_ZN3ray4core10CoreWorker14CancelChildrenERKNS_6TaskIDEb+0x149) [0x7fd24cbfd369] ray::core::CoreWorker::CancelChildren()
(task pid=279406) /path/to/my/project/.venv/lib/python3.10/site-packages/ray/_raylet.so(_ZN3ray4core10CoreWorker16HandleCancelTaskENS_3rpc17CancelTaskRequestEPNS2_15CancelTaskReplyESt8functionIFvNS_6StatusES6_IFvvEES9_EE+0x307) [0x7fd24cc82c97] ray::core::CoreWorker::HandleCancelTask()
(task pid=279406) /path/to/my/project/.venv/lib/python3.10/site-packages/ray/_raylet.so(_ZNSt17_Function_handlerIFvvEZN3ray3rpc14ServerCallImplINS2_24CoreWorkerServiceHandlerENS2_17CancelTaskRequestENS2_15CancelTaskReplyEE13HandleRequestEvEUlvE_E9_M_invokeERKSt9_Any_data+0x110) [0x7fd24cc6ec30] std::_Function_handler<>::_M_invoke()
(task pid=279406) /path/to/my/project/.venv/lib/python3.10/site-packages/ray/_raylet.so(+0x8b9826) [0x7fd24cf18826] EventTracker::RecordExecution()
(task pid=279406) /path/to/my/project/.venv/lib/python3.10/site-packages/ray/_raylet.so(+0x85c8de) [0x7fd24cebb8de] std::_Function_handler<>::_M_invoke()
(task pid=279406) /path/to/my/project/.venv/lib/python3.10/site-packages/ray/_raylet.so(+0x85cb36) [0x7fd24cebbb36] boost::asio::detail::completion_handler<>::do_complete()
(task pid=279406) /path/to/my/project/.venv/lib/python3.10/site-packages/ray/_raylet.so(+0xcf57bb) [0x7fd24d3547bb] boost::asio::detail::scheduler::do_run_one()
(task pid=279406) /path/to/my/project/.venv/lib/python3.10/site-packages/ray/_raylet.so(+0xcf69f1) [0x7fd24d3559f1] boost::asio::detail::scheduler::run()
(task pid=279406) /path/to/my/project/.venv/lib/python3.10/site-packages/ray/_raylet.so(+0xcf6c60) [0x7fd24d355c60] boost::asio::io_context::run()
(task pid=279406) /path/to/my/project/.venv/lib/python3.10/site-packages/ray/_raylet.so(_ZN3ray4core10CoreWorker12RunIOServiceEv+0xcd) [0x7fd24cc7bfed] ray::core::CoreWorker::RunIOService()
(task pid=279406) /path/to/my/project/.venv/lib/python3.10/site-packages/ray/_raylet.so(+0xe2aa10) [0x7fd24d489a10] execute_native_thread_routine
(task pid=279406) /lib/x86_64-linux-gnu/libc.so.6(+0x94b43) [0x7fd24e144b43]
(task pid=279406) /lib/x86_64-linux-gnu/libc.so.6(+0x126a00) [0x7fd24e1d6a00]
(task pid=279406) 
```

### Versions / Dependencies

Ray 2.2.0
Python 3.10.6
Ubuntu 22.04

### Reproduction script

```python
import asyncio

import ray


@ray.remote(num_cpus=0)
class Actor1:
    def __init__(self):
        self.event = asyncio.Event()

    def set(self):
        self.event.set()

    async def wait(self):
        await self.event.wait()


@ray.remote(num_cpus=0)
class Actor2:
    def __init__(self, obj):
        (self.obj,) = obj

    def cancel(self):
        ray.cancel(self.obj)


@ray.remote
def task():
    return ray.get(a1.wait.remote())


a1 = Actor1.remote()
t = task.remote()
a2 = Actor2.remote((t,))
a2.cancel.remote()
ray.get(t)
```

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by rkooo567**
Signed-off-by: SangBin Cho <rkooo567@gmail.com>

## Why are these changes needed?

`ray.cancel` is only supported for tasks, not actor tasks (https://docs.ray.io/en/master/ray-core/package-ref.html#ray-cancel). Note that it is an intended design because canceling actor tasks could corrupt the actor states easily. 

When `ray.cancel` is called, we set `recursive=True`, which means all children's tasks will also be canceled. However, when this happens, if the task has a child "actor task", it crashes the worker with `WorkerCrashedError: task_spec.cc:200: Check failed: sched_cls_id_ > 0` because we don't handle this case properly. 

To fix the issue, we check if the child tasks are actor task. This PR also improves the error message when recursive cancellation is failed. Note that because ray.cancel is not blocking, we couldn't include the error message into ray.get(canceled_task). 

## Related issue number

Closes https://github.com/ray-project/ray/issues/31398

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rkooo567** [comment]
Screen Shot 2023-02-06 at 6 21 08 PM

Error message sample. Previously, it didn't print any meaningful error messages.

File: src/ray/core_worker/core_worker.cc [hunk 1]
```diff
@@ -2212,22 +2212,53 @@ Status CoreWorker::CancelTask(const ObjectID &object_id,
...
   bool recursive_success = true;
   for (const auto &child_id : task_manager_->GetPendingChildrenTasks(task_id)) {
     auto child_spec = task_manager_->GetTaskSpec(child_id);
-    if (child_spec.has_value()) {
+    if (!child_spec.has_value()) {
+      recursive_success = false;
+      recursive_cancellation_status.push_back(
+          std::make_pair(child_id,
+                         Status::UnknownError(
+                             "Recursive task cancellation failed--check warning logs.")));
```
Comment (rickyyx) [inline src/ray/core_worker/core_worker.cc:2224]:
```suggestion
                             "Recursive task cancellation failed. Check warning logs.")));
```

File: src/ray/core_worker/core_worker.cc [hunk 2]
```diff
@@ -2212,22 +2212,53 @@ Status CoreWorker::CancelTask(const ObjectID &object_id,
...
+    } else if (child_spec->IsActorTask()) {
+      recursive_success = false;
+      recursive_cancellation_status.push_back(std::make_pair(
+          child_id,
+          Status::Invalid(
+              "Actor task cancellation is not supported. The task won't be cancelled.")));
```
Comment (rickyyx) [inline src/ray/core_worker/core_worker.cc:2230]: wondering if we should allow this in the future.

File: src/ray/core_worker/core_worker.cc [hunk 3]
```diff
@@ -2212,22 +2212,53 @@ Status CoreWorker::CancelTask(const ObjectID &object_id,
...
+      recursive_cancellation_status.push_back(std::make_pair(child_id, result));
     }
   }
+
   if (recursive_success) {
     return Status::OK();
   } else {
-    return Status::UnknownError(
-        "Recursive task cancellation failed--check warning logs.");
+    auto kMaxFailedTaskSampleSize = 10;
```
Comment (rickyyx) [inline src/ray/core_worker/core_worker.cc:2241]: Do we have tests cover this?

**Review by scv119 — APPROVED** [review]

File: src/ray/core_worker/core_worker.cc (hunk 1 above)
Comment (rkooo567) [reply src/ray/core_worker/core_worker.cc:2224]: lol surprisingly, this is just copied and pasted. But it makes sense to change it

File: src/ray/core_worker/core_worker.cc (hunk 3 above)
Comment (rkooo567) [reply src/ray/core_worker/core_worker.cc:2241]: added one now

**Pull request merged**